crates/sparrow-journal/src/batch_io.rs

```diff
@@ -0,0 +1,216 @@
+    }
+}
+
+pub(super) struct Encoded {
+    up_to_time: RowTime,
+    dictionary_batches: Vec<EncodedData>,
+    batch: Option<EncodedData>,
+}
+
+impl Encoded {
+    pub(super) fn write_to_journal(
+        self,
+        wal: &mut okaywal::WriteAheadLog,
+    ) -> std::io::Result<okaywal::EntryId> {
+        let mut entry = wal.begin_entry()?;
+        entry.write_chunk(&self.up_to_time.bytes())?;
+        for dictionary_batch in self.dictionary_batches {
+            entry.write_chunk(&dictionary_batch.ipc_message)?;
+            entry.write_chunk(&dictionary_batch.arrow_data)?;
+        }
+        if let Some(batch) = self.batch {
+            entry.write_chunk(&batch.ipc_message)?;
+            entry.write_chunk(&batch.arrow_data)?;
+        }
+        entry.commit()
+    }
+}
+
+impl BatchEncoder {
+    pub fn new(schema: SchemaRef) -> error_stack::Result<Self, Error> {
+        Ok(Self {
+            dictionary_tracker: DictionaryTracker::new(false),
+            gen: IpcDataGenerator {},
+            write_options: IpcWriteOptions::default(),
+            schema,
+        })
+    }
+
+    pub fn encode(&mut self, batch: &Batch) -> error_stack::Result<Encoded, Error> {
+        // TODO: Split batches to fit in a certain size.
+        let up_to_time = batch.up_to_time;
+
+        // Create a record batch for the given batch.
+        if let Some(batch) = &batch.data {
+            let batch = RecordBatch::try_new(
+                self.schema.clone(),
+                vec![
+                    batch.time.clone(),
+                    batch.subsort.clone(),
+                    batch.key_hash.clone(),
+                    batch.data.clone(),
+                ],
+            )
+            .into_report()
+            .change_context(Error::WriteBatch)?;
+            let (dictionary_batches, batch) = self
+                .gen
+                .encoded_batch(&batch, &mut self.dictionary_tracker, &self.write_options)
+                .into_report()
+                .change_context(Error::WriteBatch)?;
+            Ok(Encoded {
+                up_to_time,
+                dictionary_batches,
+                batch: Some(batch),
+            })
+        } else {
+            Ok(Encoded {
+                up_to_time,
+                dictionary_batches: vec![],
+                batch: None,
+            })
+        }
+    }
+}
+
+#[derive(Debug)]
+pub(super) struct BatchDecoder {
+    schema: SchemaRef,
+
+    /// Optional dictionaries for each schema field.
+    ///
+    /// Dictionaries may be appended to in the streaming format.
+    dictionaries_by_id: HashMap<i64, ArrayRef>,
+}
+
+impl BatchDecoder {
+    pub fn new(schema: SchemaRef) -> Self {
+        Self {
+            schema,
+            dictionaries_by_id: HashMap::new(),
+        }
+    }
+
+    /// Decode the batch from the given entry.
+    ///
+    /// If it was completely written, will return `Some(batch)`. If it wasn't completely
+    /// written, will return `None`.
```

Comment:
Is `None` a valid state?  What does it mean to not completely write a batch? I didn't see any further explanation in okayWal.

Reply:
https://khonsulabs.github.io/okaywal/main/okaywal/struct.Entry.html#method.read_chunk